### ISSUE TITLE
Legion: Better Error Handling

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#632]
This is my omnibus issue for better error handling. Many of these items rely on the same fundamental changes inside the runtime and therefore I am keeping them together as a single issue. I expect that all these fixes will land at once and not independently, which justifies handling them together in a single issue. Feel free to append to the list.

- [ ] Use names of regions/fields/tasks that have been registered in error messages
- [ ] Provide a task stack trace for errors
- [ ] Decouple runtime error checking and messages from debug mode
- [ ] Switch to global UniqueIDs so that we can always uniquely recognize operations
- [ ] Use abort instead of assert when exiting from errors
- [ ] Sort the error IDs into a scheme that makes sense
- [ ] Warn as error mode
- [ ] Provide better errors for typed operations like dependent partitioning operations
- [ ] Have a clean shutdown methodology after detecting an error so clean-up can happen
- [ ] Hooks for capturing file, line number, or backtrace of call sites for runtime calls
- [ ] #426

**Comment by lightsighter** [comment]
Remember to make use of the new semantic tags from #819

**Comment by elliottslaughter** [comment]
When we're doing this, it would be nice to use not just the names directly attached to a region, but also names attached to parent/ancestor regions as well. E.g. if you have a region tree `R -> P -> S` with a name `foo` attached to `R` but no names attached to either `P` or `S`, it would be nice to get a default-generated name like `foo.part(i)[j]` (where `i` and `j` are the colors of `P` and `S`, respectively) rather than not getting any name at all.

**Comment by aheirich** [comment]
I would like to add to this list, I assume this is obvious:
Change all runtime asserts in debug node to error messages


On Apr 9, 2020, at 2:09 PM, Elliott Slaughter <notifications@github.com<mailto:notifications@github.com>> wrote:



When we're doing this, it would be nice to use not just the names directly attached to a region, but also names attached to parent/ancestor regions as well. E.g. if you have a region tree R -> P -> S with a name foo attached to R but no names attached to either P or S, it would be nice to get a default-generated name like foo.part(i)[j] (where i and j are the colors of P and S, respectively) rather than not getting any name at all.

**Comment by lightsighter** [comment]
@elliottslaughter What happens if there are no "named" regions anywhere in the tree? What do you think should happen?

@aheirich There will be three modes of checking after I'm done with this issue (see the third bullet):
1. Runtime will full safety checks: will check all error conditions regardless of how you compiled the runtime, errors will be reported and the runtime will exit, can be toggled with a command line flag
2. Runtime with no safety checks: will not check any error conditions regardless of how you compiled the runtime, can be toggled with a command line flag
3. Runtime debug mode: which will abort on errors so you can see backtraces when you hit them, for runtime debugging only

**Comment by elliottslaughter** [comment]
I guess if you don't have any names at all (for the node in question and its ancestors) then you'd fall back to using something like `unnamed` or `unknown` for the name. I do think we should continue to give the region tree ID and so on (even in the case where the region is named), so in practice the error would look like:

```
Error: task asdf does not have privileges on field "unnamed" (12) of region "unnamed" (1, 2, 3) ...
```

The IDs are useful even in the named case, in case the user creates two regions with the same name.

**Comment by lightsighter** [comment]
Added bullet for recording that IDs use decimal and not hex values.

**Comment by lightsighter** [comment]
Update type encoding errors per #851

**Comment by aheirich** [comment]
I'm having a runtime problem using FieldAccessors
LEGION ERROR: Accessor construction for field 3 in task copyGammaTableTask without privileges!
The construction is here
  const FieldAccessor<WRITE_DISCARD, double, 2, coord_t,
        Realm::AffineAccessor<double, 2, coord_t> >
        acc(regions[0], GAMMA_TABLE_FIELD_ID);
The region requirement for the task specified WRITE_DISCARD
  RegionRequirement gammaRequirement(gamma_table_lr, WRITE_DISCARD, EXCLUSIVE, gamma_table_lr);

lightsighter:rocket:  10 days ago
The problem is that field doesn't exist in that region requirement so you don't have privileges on it. It's the reason the error message says exactly what field is bad. If you messed up the kind of privilege then you would get a different error message.

aheirich  9 days ago
kind of a confusing message, would you agree?

lightsighter:rocket:  9 days ago
i think it says exactly what it means, you don't have privileges on that field

lightsighter:rocket:  9 days ago
therefore you can't make an accessor for it

aheirich  8 days ago
The problem turned out to be that I forgot to add_field().  So it was the field that was missing, not the privilege.  The privilege was specified in then RegionRequirement.

lightsighter:rocket:  8 days ago
That's exactly what that message means. If you didn't include the field in the region requirement then you by definition do not have privileges on it.

aheirich  7 days ago
It's misleading for tthe message to mention privileges

lightsighter:rocket:  7 days ago
How do you think the runtime should distinguish this case from the field being created in a sub-task and not returned back to the parent task? Both of them will manifest the same way and you're asking for a special treatment for the error message on the first case, so I need a decision procedure to recognize it.

aheirich  7 days ago
it sounds like the field is missing in both cases, not the privilege

lightsighter:rocket:  7 days ago
It's possible to have the field on the region requirement but the context to not have privileges, I can't tell the difference with the meta-data that I have at this point in the code

lightsighter:rocket:  7 days ago
And I can't add that meta-data without incurring overhead

aheirich  7 days ago
maybe the message should say "missing privilege or missing field, I can't tell which"

**Comment by aheirich** [comment]
Mike, can you help me understand how this statement
348 typedef const FieldAccessor<READ_WRITE, ImageReduction::PixelField,
349 image_region_dimensions,
350 coord_t, Realm::AffineAccessor<ImageReduction::PixelField,
351 image_region_dimensions, coord_t> > RWAccessor;
352 __TRACE
353 RWAccessor r0In(image0_pr, ImageReduction::FID_FIELD_R);
can generate this error message? the error is from line 353
[0 - 7fb23f928700] {5}{runtime}: [error 164] LEGION ERROR: Dynamic type mismatch in 'get_index_space_domain' (from file /home/aheirich/Documents/Legion/software/legion/runtime/legion/region_tree.inl:1950)[error 164]

aheirich 11:08 AM
I found it - the physical region had been created with a 2D index space but the accessor was 3D.
In retrospect I can understand this message, but when I first read it the word "type" made me think the problem was the data type. This is confusing.

lightsighter🚀 3:52 PM
legion can't tell the difference in realm's type encodings, all it has is an in integer and we don't know whether that integer is different because the coordinate type is different or the dimensions are different

aheirich 3:53 PM
then could the error message say that? that would be more helpful
New

lightsighter🚀 3:55 PM
i think you should just start making notes on this on the github issue for improved error messages, i'm not going to be going through and changing lots of error messages that will create a bunch of unnecessary conflicts in other branches that are outstanding right now

**Comment by lightsighter** [comment]
Need to check all user-specified values that could be AUTO_GENERATE_ID to make sure the user has given us a good value. For example, we need to check that partition colors do not conflict with previously specified partition colors for the same parent index space.

**Comment by aheirich** [comment]
I am getting this error
LEGION ERROR: ProjectionID 99 has already been used in the region projection table
(from file /home/groups/aaiken/seshu/legion_s3d_visualization/legion/runtime/legion/runtime.cc:16204)
although there are no calls to register_region_function there is a call to register_projection_functor for ID 99.  Could this error message be confused?
white_check_mark
eyes
raised_hands

7:26
yes I think the message is in error I will make a note on the relevant github issue

Here is a mistake.  This error message is from register_projection_functor but it talks about region projection table, probably a cut and paste error.  Change this to talk about the projection functor id table instead.

**Comment by lightsighter** [comment]
I agree that the above error message should not contain the word "region". Everything else about it is both sound and accurate. You cannot record multiple projection functions with the same projection ID.
